controllers/complianceeventsapi/types.go

```diff
@@ -0,0 +1,266 @@
+package complianceeventsapi
+
+import (
+	"bytes"
+	"crypto/sha1"
+	"database/sql/driver"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/lib/pq"
+)
+
+var (
+	errRequiredFieldNotProvided = errors.New("required field not provided")
+	errInvalidInput             = errors.New("invalid input")
+)
+
+type ComplianceEvent struct {
+	Cluster      Cluster       `json:"cluster"`
+	Event        EventDetails  `json:"event"`
+	ParentPolicy *ParentPolicy `json:"parent_policy,omitempty"` //nolint:tagliatelle
+	Policy       Policy        `json:"policy"`
+}
+
+func (ce ComplianceEvent) Validate() error {
+	errs := make([]error, 0)
+
+	if err := ce.Cluster.Validate(); err != nil {
+		errs = append(errs, err)
+	}
+
+	if ce.ParentPolicy != nil {
+		if err := ce.ParentPolicy.Validate(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if err := ce.Event.Validate(); err != nil {
+		errs = append(errs, err)
+	}
+
+	if err := ce.Policy.Validate(); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errors.Join(errs...)
+}
+
+type Cluster struct {
+	KeyID     int    `db:"id" json:"-" goqu:"skipinsert"`
+	Name      string `db:"name" json:"name"`
+	ClusterID string `db:"cluster_id" json:"cluster_id"` //nolint:tagliatelle
+}
+
+func (c Cluster) Validate() error {
+	errs := make([]error, 0)
+
+	if c.Name == "" {
+		errs = append(errs, fmt.Errorf("%w: cluster.name", errRequiredFieldNotProvided))
+	}
+
+	if c.ClusterID == "" {
+		errs = append(errs, fmt.Errorf("%w: cluster.cluster_id", errRequiredFieldNotProvided))
+	}
+
+	return errors.Join(errs...)
+}
+
+type EventDetails struct {
+	KeyID      int       `db:"id" json:"-" goqu:"skipinsert"`
+	ClusterID  int       `db:"cluster_id" json:"-"`
+	PolicyID   int       `db:"policy_id" json:"-"`
+	Compliance string    `db:"compliance" json:"compliance"`
+	Message    string    `db:"message" json:"message"`
+	Timestamp  time.Time `db:"timestamp" json:"timestamp"`
+	Metadata   JSONMap   `db:"metadata" json:"metadata,omitempty"`
+	ReportedBy *string   `db:"reported_by" json:"reported_by,omitempty"` //nolint:tagliatelle
+}
+
+func (e EventDetails) Validate() error {
+	errs := make([]error, 0)
+
+	if e.Compliance == "" {
+		errs = append(errs, fmt.Errorf("%w: event.compliance", errRequiredFieldNotProvided))
+	} else {
+		switch e.Compliance {
+		case "Compliant", "compliant", "NonCompliant", "noncompliant":
```

Comment:
What do you think about just allowing `compliant` and `noncompliant` so we don't have to account for different cases in queries later on?

Reply:
I agree in principle, but I think our resources most often use Compliant and NonCompliant, so that's what I'll make this accept. Unless you have a really good reason to use all lowercase?

Reply:
@JustinKuli I think just allowing a single variation of the value is what I'm getting at. If we want to use `Compliant` and `NonCompliant`, that sounds good to me, though `NonCompliant` is not correct English and is strange for non-OCM use-cases if this gets reused. :laughing: